### PR TITLE
[Elastic Agent]-PR- Introduce log message for not supported annotations for Hints based autodiscover

### PR DESCRIPTION
**Pull request by gizas**
<del>Note: Dont merge  until https://github.com/elastic/elastic-agent-autodiscover/pull/81 is merged.</del>
<del>We need to udpate go.mod of the elastic-agent with v0.6.9</del> PR is merged

- Enhancement

## What does this PR do?

Relates to:
- https://github.com/elastic/elastic-agent/issues/3063
- https://github.com/elastic/elastic-agent/issues/3064
- https://github.com/elastic/elastic-agent-autodiscover/pull/81

## Why is it important?

It adds a helpful feature for those that use Hints Autodiscovery. It can be helpful to identify configuration issues

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works (Some tests updated [here](https://github.com/elastic/elastic-agent-autodiscover/pull/81/files#diff-2f7d9a00d75cb68d90813219822383fdda49c60b39cc05f7d2713f3d9e18a84a))
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## How to test this PR locally

Follow same instructions as https://github.com/elastic/elastic-agent/pull/3107
1. Checkout this branch
2. Update hints.go file of your <home>/go/pkg/mod/github.com/elastic/elastic-agent-autodiscover@v0.6.8/utils as per [hints.go](https://github.com/elastic/elastic-agent-autodiscover/pull/81/files#diff-99d5db949c88160c717cd1240bed90ec4da0d96b04cd287de3ca53d0cd90e4bf)
3. Build elastic agent image
```bash
EXTERNAL=true DEV=true SNAPSHOT=true PLATFORMS=linux/arm64 PACKAGES=docker mage package
cd build/package/elastic-agent/elastic-agent-linux-arm64.docker/docker-build
docker build -t custom-agent-image .\n
kind load docker-image custom-agent-image:latest\n
```
4. Install elastic-agent-standalone with hints enabled. Sample here
[elastic-agent-standalone.txt](https://github.com/elastic/elastic-agent/files/14509781/elastic-agent-standalone.txt)

5. Configure some "wrong" co.elastic.hints annotation inside a pod. 
```yaml
annotations:        
        co.elastic.hints/enabled: 'true'
        co.elastic.hints/package: "container_logs"
        co.elastic.hints/hostssssss: '128.30.40.50'
```

7. Warning messages should be available 

## Related issues

- Relates https://github.com/elastic/elastic-agent/issues/3064

## Logs

Deploy following pod with wrong annotations hosts and steam:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nginx
  name: nginx
  namespace: default
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        co.elastic.hints/enabled: 'true'
        co.elastic.hints/hosts: '128.30.40.50'
        co.elastic.hints/steam: 'stdout'
    spec:
      containers:
      - image: nginx
        name: nginx
```

```bash
{"log.level":"warn","@timestamp":"2024-03-07T13:15:22.599Z","log.logger":"composable.providers.kubernetes","log.origin":{"file.name":"kubernetes/pod.go","file.line":219},"message":"provided hint: co.elastic/hints/hosts is not in the supported list for pod nginx-68d79864cb-rgjbb in namespace default","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2024-03-07T13:15:22.599Z","log.logger":"composable.providers.kubernetes","log.origin":{"file.name":"kubernetes/pod.go","file.line":219},"message":"provided hint: co.elastic/hints/steam is not in the supported list for pod nginx-68d79864cb-rgjbb in namespace default","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
```

Check that only annotations that start with co.elastic are the ones that are checked for correctness.

## Screenshots

Screenshot 2024-03-07 at 3 21 19 PM

### CODE REVIEW

**Comment by gizas** [comment]
1. I might need to add integration tests
2. To update changelog when is done
3. The for loops is also present in elastic-agent-autodiscover library. Wondering if I can put there the block of checks to minmimise the for loops that happen. Now we add iterate again to take annotations of pods

File: internal/pkg/composable/providers/kubernetes/pod.go
Comment (gizas) [inline internal/pkg/composable/providers/kubernetes/pod.go:218]: At the moment I set as warning.
This kind of errors I think must be visible with default installation

But almost every 2sec we have a warning for one pod per wrong annotation. This can add noise
What do you think?

Comment (MichaelKatsoulis) [reply internal/pkg/composable/providers/kubernetes/pod.go:218]: They will have it every two seconds until they fix it. If the only reason to get this warning is a wrong hint annotation, this will force them to fix it. 
I would leave it as a warning.

**Comment by AndersonQ** [comment]
@gizas could you make it a draft until it's ready for review?

File: internal/pkg/composable/providers/kubernetes/hints.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints.go:33]: why we need here `"enabled", "module"` ? it is not expected to use `module` for the elastic-agent, isn't it?

File: internal/pkg/composable/providers/kubernetes/hints.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints.go:33]: `"package"` is not needed - it is defined in the list of `const` here - https://github.com/elastic/elastic-agent/pull/4360/files#diff-62d5eda37f335c34852c11b28a79d8f6c3010449cdd5a3a8c91d1fb4c8bbf000R21

**Comment by gizas** [comment]
Converted to Draft as more tests need to be covered

File: go.mod
Comment (cmacknz) [inline go.mod:16]:
```suggestion
	github.com/elastic/elastic-agent-autodiscover v0.6.10
```

There's a newer version now updating the protobuf version to avoid a CVE.

Comment (gizas) [reply go.mod:16]: v0.6.11 on the way that included v0.6.10

File: internal/pkg/composable/providers/kubernetes/hints.go
Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints.go:33]: It is for the actual hints where there can be explicitly enabled per container
Like the following:
```yaml
annotations:
        co.elastic.hints/enabled: 'true'
        co.elastic.hints/hosts: '128.30.40.50'
        co.elastic.hints/streams: 'error'
```

File: go.mod
Comment (gizas) [reply go.mod]: updated to v0.6.12

File: internal/pkg/composable/providers/kubernetes/pod.go
Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/pod.go]: Done

**Comment by gizas** [comment]
@elastic/elastic-agent-data-plane can I have a review please?

File: internal/pkg/composable/providers/kubernetes/hints.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints.go:33]:
```suggestion
var allSupportedHints = []string{"enabled", integration, datastreams, host, period, timeout, metricspath, username, password, stream, processors}
```

`"package"` is already defined https://github.com/elastic/elastic-agent/pull/4360/files#diff-62d5eda37f335c34852c11b28a79d8f6c3010449cdd5a3a8c91d1fb4c8bbf000R21

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints_test.go:379]: is the different format of labels relevant for this test?

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints_test.go:419]: what is the typo here?

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints_test.go]: No, copy paste from other test

I removed them in latest commit

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints_test.go:415]: Should be metrics_path

https://github.com/elastic/elastic-agent/blob/5a5e0f3f79a4b69c32762928417861c1ed033280/internal/pkg/composable/providers/kubernetes/hints.go#L26C17-L26C29

Comment (tetianakravchenko) [reply internal/pkg/composable/providers/kubernetes/hints_test.go:415]: for the purpose of the test - it is not important either it is a typo or not supported hint, maybe to make it obvious can be used smth like `"notSupportedHint"`?

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints_test.go:444]: can it be removed as well? the same in `pod_test`

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (tetianakravchenko) [inline internal/pkg/composable/providers/kubernetes/hints_test.go:504]: containerID `asdfghjkl` is not needed here? there is a dedicated test for this

Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints_test.go]: Done in [62d1b2d](https://github.com/elastic/elastic-agent/pull/4360/commits/62d1b2d7af52bfe09406cff957b6be838b147d1c)

File: internal/pkg/composable/providers/kubernetes/hints_test.go
Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints_test.go]: Done thanks!

Comment (gizas) [reply internal/pkg/composable/providers/kubernetes/hints_test.go]: Done in [62d1b2d](https://github.com/elastic/elastic-agent/pull/4360/commits/62d1b2d7af52bfe09406cff957b6be838b147d1c)

**Review by tetianakravchenko — APPROVED** [review]

**Review by constanca-m — APPROVED** [review]

**Comment by gizas** [comment]
@AndersonQ , @pchila (sorry for direct ping as you were assigned to issue) can I have a review here please?
Also can you see why this is failing to build agent? 

See my comment before https://github.com/elastic/elastic-agent/pull/4360#issuecomment-2069606919

File: go.mod
Comment (blakerouse) [inline go.mod:14]: This should not be here. Happens sometimes when the build directory has extracted beats in it. This needs to be cleaned up before merging.

**Review by blakerouse — CHANGES_REQUESTED** [review]

File: go.mod
Comment (gizas) [reply go.mod]: Thanks @blakerouse, I cleaned /bin folder and made the update.
Seems that not got passed and check-ci does not fail

(The failures are not related with this changes I think)

Comment (gizas) [reply go.mod]: @blakerouse I have tried to re-run the build with no luck. 
I can see lots of `ResourceExistsError `

Also for the SonarQube Analysis do you have a suggestion?

**Comment by gizas** [comment]
@blakerouse , @AndersonQ any ideas on the above errors?

**Comment by pchila** [comment]
> @blakerouse , @AndersonQ any ideas on the above errors?

Check CI is now passing, Ci run hit a flaky test on windows tracked already by https://github.com/elastic/elastic-agent/issues/4336 

As for the serverless tests it seems you have an unexpected error log `add_cloud_metadata: received error failed fetching EC2 Identity Document: not found, Signing` (it seems it started failing in the same way on main as well) while the extend runtime leak tests just failed for an orchestration issue...

**Comment by gizas** [comment]
@pchila indeed the check ci is passing since the changes [here](https://github.com/elastic/elastic-agent/pull/4360#discussion_r1579402313)

Are you aware about any fix for the rest of tests?
For sonarcube analysis any idea?

**Comment by pchila** [comment]
> @pchila indeed the check ci is passing since the changes [here](https://github.com/elastic/elastic-agent/pull/4360#discussion_r1579402313) 
> 
> Are you aware about any fix for the rest of tests? For sonarcube analysis any idea?

@gizas the integration tests failures should be fixed by https://github.com/elastic/elastic-agent/pull/4632 (already merged on main so you need to update your branch)

For the sonarqube analysis it seems that your new code is not covered very well by unit tests...

**Comment by gizas** [comment]
@pchila thank you for the inputs here. 

I merged with main but still elastic-agent build fails. 
I see `>> (linux-arm64-ubuntu-2204-default) Test output (non-sudo) (stdout): === FAIL: testing/integration TestOtelFileProcessing (188.30s)` is the one that always fails

Moreover for the Sonarcube coverage, my additions are made in private [emitRunning function](https://github.com/elastic/elastic-agent/pull/4360/files#diff-ed933ad9cf09eb9ca797c5c89b5e8991624cbe5301d88f996d9bd947f4e3ea16R220) which is called on Add event. This fucntion was not covered even before introduction of SonarCube

I added a test [here](https://github.com/elastic/elastic-agent/pull/4360/files#diff-0f5944c93f2d789680477962de5fd0fdd1ea41e41c8593aa6542182a6204dcf1R379) but obviously is not matched with specific function. Can we merge this despite the sonarcube failure?

**Comment by pchila** [comment]
> @pchila thank you for the inputs here.
> 
> I merged with main but still elastic-agent build fails. I see `>> (linux-arm64-ubuntu-2204-default) Test output (non-sudo) (stdout): === FAIL: testing/integration TestOtelFileProcessing (188.30s)` is the one that always fails
> 

This is tracked by https://github.com/elastic/elastic-agent/issues/4617 as it shows up on main as well.

> Moreover for the Sonarcube coverage, my additions are made in private [emitRunning function](https://github.com/elastic/elastic-agent/pull/4360/files#diff-ed933ad9cf09eb9ca797c5c89b5e8991624cbe5301d88f996d9bd947f4e3ea16R220) which is called on Add event. This fucntion was not covered even before introduction of SonarCube
> 
> I added a test [here](https://github.com/elastic/elastic-agent/pull/4360/files#diff-0f5944c93f2d789680477962de5fd0fdd1ea41e41c8593aa6542182a6204dcf1R379) but obviously is not matched with specific function. Can we merge this despite the sonarcube failure?

Without having a green sonarqube check, someone with admin privileges should force the merge
I didn't have the time to fully review your merge yet but I would really prefer if we could cover new code/functionality with unit tests rather than force-merge without automated unit tests

**Comment by gizas** [comment]
Nice!!!! it seems all tests are successful !!!!
@blakerouse , @pchila can I have final review and merge if no objection? (I have enabled auto-merge)

**Comment by cmacknz** [comment]
@blakerouse please re-review, looks like go.mod has been fixed.

**Comment by cmacknz** [comment]
buildkite test it

**Comment by gizas** [comment]
@blakerouse please any objection before merging?

**Review by blakerouse — APPROVED** [review]
Looks good.

**Pull request merged**
